src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+        _childModified |= serverModified;
+    }
+    if (localEntry.isValid()) {
+        item->_inode = localEntry.inode;
+        bool typeChange = dbEntry.isValid() && localEntry.isDirectory != (dbEntry._type == ItemTypeDirectory);
+        if (dbEntry.isValid() && localEntry.isVirtualFile) {
+            if (dbEntry._type == ItemTypeFile) {
+                // If we find what looks to be a spurious "abc.owncloud" the base file "abc"
+                // might have been renamed to that. Make sure that the base file is not
+                // deleted from the server.
+                if (dbEntry._modtime == localEntry.modtime && dbEntry._fileSize == localEntry.size) {
+                    qCInfo(lcDisco) << "Base file was renamed to virtual file:" << item->_file;
+                    Q_ASSERT(item->_file.endsWith(_discoveryData->_syncOptions._virtualFileSuffix));
+                    item->_direction = SyncFileItem::Down;
+                    item->_instruction = CSYNC_INSTRUCTION_NEW;
+                    item->_type = ItemTypeVirtualFile;
+                }
+            } else if (item->_type != ItemTypeVirtualFileDownload) {
+                item->_type = ItemTypeVirtualFile;
+            }
+
+            if (noServerEntry) {
+                item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+                item->_direction = SyncFileItem::Down;
+            }
+        } else if (!dbEntry.isValid() && localEntry.isVirtualFile && !noServerEntry) {
+            // Somehow there is a missing DB entry while the virtual file already exists.
+            // The instruction should already be set correctly.
+            ASSERT(item->_instruction == CSYNC_INSTRUCTION_UPDATE_METADATA);
+            ASSERT(item->_type == ItemTypeVirtualFile)
+            ASSERT(item->_file.endsWith(_discoveryData->_syncOptions._virtualFileSuffix));
+            item->_file.chop(_discoveryData->_syncOptions._virtualFileSuffix.size());
+        } else if (dbEntry.isValid() && !typeChange && ((dbEntry._modtime == localEntry.modtime && dbEntry._fileSize == localEntry.size) || (localEntry.isDirectory && dbEntry._type == ItemTypeDirectory))) {
+            // Local file unchanged.
+            if (noServerEntry) {
+                item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+                item->_direction = SyncFileItem::Down;
+            } else if (!serverModified && dbEntry._inode != localEntry.inode) {
+                item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+                item->_direction = SyncFileItem::Down; // Does not matter
+            }
+        } else if (serverModified || dbEntry._type == ItemTypeVirtualFile) {
+            if (serverEntry.isDirectory && localEntry.isDirectory) {
+                // Folders of the same path are always considered equals
+                item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+            } else {
+                // It could be a conflict even if size and mtime match!
```

Comment:
Wait! If local/db size+mtime match, we won't be in this branch. We'll have gone into the above `else if`!

Possibly this comment is confusing here since the below code seems to assume that there was a local change.

Reply:
Right, this comment is copy-pasted from the reconcile code, but some code was moved around and is no longer in the same order.